File: app/src/main/java/uk/co/transferx/app/signup/fragment/SignUpStepTwoFragment.java
package uk.co.transferx.app.signup.fragment;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.TextInputEditText;
import android.text.Editable;
import android.text.InputType;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import javax.inject.Inject;

import uk.co.transferx.app.BaseFragment;
import uk.co.transferx.app.R;
import uk.co.transferx.app.TransferXApplication;
import uk.co.transferx.app.signup.SignUpActivity;
import uk.co.transferx.app.signup.presenters.SignUpStepTwoPresenter;
import uk.co.transferx.app.welcom.WelcomeActivity;

import static uk.co.transferx.app.util.Constants.EMAIL;
import static uk.co.transferx.app.util.Constants.PASSWORD;
import static uk.co.transferx.app.util.Constants.U_NAME;

/**
 * Created by smilevkiy on 15.11.17.
 */

public class SignUpStepTwoFragment extends BaseFragment implements SignUpStepTwoPresenter.SignUpStepTwoUI {

    @Inject
    SignUpStepTwoPresenter presenter;
    private TextInputEditText firstInput, secondInput;
    private TextView firstLabel, secondLabel, firstError, secondError;
    private TextWatcher firstTextWatcher, secondTextWatcher;
    private ImageView backButton;

    @Override
    public String tagName() {
        return SignUpStepTwoFragment.class.getSimpleName();
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        ((TransferXApplication) getActivity().getApplication()).getAppComponent().inject(this);
        super.onCreate(savedInstanceState);
        Bundle bundle = getArguments();
        if (bundle != null) {
            presenter.setName(bundle.getString(U_NAME));
        }
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.sign_up_step_fragment_layout, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        ((TextView) view.findViewById(R.id.title)).setText(getString(R.string.title_two));
        firstInput = view.findViewById(R.id.first_input);
        secondInput = view.findViewById(R.id.second_input);
        firstLabel = view.findViewById(R.id.first_input_label);
        secondLabel = view.findViewById(R.id.second_input_label);
        firstError = view.findViewById(R.id.first_input_error);
        secondError = view.findViewById(R.id.second_input_error);
        backButton = view.findViewById(R.id.back_button);
        backButton.setOnClickListener(v -> {
            hideKeyboard(backButton);
            getActivity().onBackPressed();
        });
        firstLabel.setText(R.string.email_address);
        secondLabel.setText(R.string.password);
        firstInput.setHint(R.string.enter_your_email_address);
        secondInput.setHint(R.string.enter_your_password);
        firstInput.setInputType(InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_EMAIL_ADDRESS);
        secondInput.setInputType(InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_PASSWORD);
        firstInput.setCompoundDrawablesWithIntrinsicBounds(R.drawable.ic_email_address, 0, 0, 0);
        secondInput.setCompoundDrawablesWithIntrinsicBounds(R.drawable.ic_password, 0, 0, 0);
        buttonNext = view.findViewById(R.id.sign_in);
        buttonNext.setOnClickListener(v -> presenter.goToNextStep());
        if (savedInstanceState != null) {
            firstInput.setText(savedInstanceState.getString(EMAIL));
            secondInput.setText(savedInstanceState.getString(PASSWORD));
        }
    }

    @Override
    public void showErrorEmail() {
        setStatusOfError(firstInput, firstLabel, R.color.red);
        firstError.setText(R.string.email_error);
        firstError.setVisibility(View.VISIBLE);
    }

    @Override
    public void showErrorPassword() {
        setStatusOfError(secondInput, secondLabel, R.color.red);
        secondError.setText(R.string.password_error);
        secondError.setVisibility(View.VISIBLE);
    }

    @Override
    public void goToNextView(String uname, String email, String password) {
        hideKeyboard(secondInput);
        Bundle bundle = new Bundle();
        bundle.putString(U_NAME, uname);
        bundle.putString(EMAIL, email);
        bundle.putString(PASSWORD, password);
        ((SignUpActivity) getActivity()).showNextOrPreviousFragment(2, bundle);
    }

    @Override
    public void onResume() {
        super.onResume();
        presenter.attachUI(this);
        presenter.setEmail(firstInput.getText().toString());
        presenter.setPassword(secondInput.getText().toString());
        firstInput.addTextChangedListener(firstTextWatcher = new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                presenter.setEmail(charSequence.toString());
            }

            @Override
            public void afterTextChanged(Editable editable) {

            }
        });
        secondInput.addTextChangedListener(secondTextWatcher = new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                presenter.setPassword(charSequence.toString());
            }

            @Override
            public void afterTextChanged(Editable editable) {

            }
        });
    }

    @Override
    public void onPause() {
        presenter.detachUI();
        firstInput.removeTextChangedListener(firstTextWatcher);
        secondInput.removeTextChangedListener(secondTextWatcher);
        super.onPause();
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        outState.putString(EMAIL, firstInput.getText().toString());
        outState.putString(PASSWORD, secondInput.getText().toString());
        super.onSaveInstanceState(outState);
    }

    @Override
    public void setStateButton(boolean isEnabled) {
        setButtonStatus(isEnabled);
    }

    @Override
    public void goToWelcome() {
       //no op
    }
}

File: app/src/main/java/uk/co/transferx/app/ui/signup/fragment/SignUpStepTwoFragment.kt
package uk.co.transferx.app.ui.signup.fragment

import android.graphics.Color
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v4.content.ContextCompat
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ProgressBar
import kotlinx.android.synthetic.main.signin_fragment_layout.*
import kotlinx.android.synthetic.main.signup_step_one_fragment_layout.*
import uk.co.transferx.app.R
import uk.co.transferx.app.TransferXApplication
import uk.co.transferx.app.ui.base.BaseFragment
import uk.co.transferx.app.ui.signup.SignUpActivity
import uk.co.transferx.app.ui.signup.presenters.SignUpStepTwoPresenter
import uk.co.transferx.app.util.Constants.*
import javax.inject.Inject

/**
 * Created by smilevkiy on 15.11.17.
 */

class SignUpStepTwoFragment : BaseFragment(), SignUpStepTwoPresenter.SignUpStepTwoUI {
    @Inject
    lateinit var presenter: SignUpStepTwoPresenter

    private var isPasswordErrorShown: Boolean = false

    private var isEmailErrorShown: Boolean = false

    private var loadingBar: ProgressBar? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        (activity?.application as TransferXApplication).appComponent.inject(this)
        super.onCreate(savedInstanceState)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.signup_step_one_fragment_layout, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        buttonBack.setOnClickListener {
            hideKeyboard(buttonBack)
            activity?.onBackPressed()
        }

        // Set credentials from previous step
        presenter.firstName = arguments?.getString(FIRST_NAME)
        presenter.lastName = arguments?.getString(LAST_NAME)
        presenter.phoneNumber = arguments?.getString(PHONE_NUMBER)
        presenter.country = arguments?.getString(COUNTRY)

        emailInputText.setOnFocusChangeListener { text, b ->
            if (!b) presenter.checkIfEmailIsTaken()
        }

        // View from activity has to be obtained in normal way
        loadingBar = activity?.findViewById<ProgressBar>(R.id.loading_bar)

        buttonNext = buttonNextStep
        buttonNext.setOnClickListener {
            loadingBar?.visibility = View.VISIBLE
            presenter.signUpUser()
            hideKeyboard()
        }

        setUpCredentialsOnRestore(savedInstanceState)
    }

    override fun goToPinSetup() {
        hideKeyboard()

        (activity as SignUpActivity).showNextOrPreviousFragment(2, null)

        loadingBar?.visibility = View.GONE
    }

    override fun onResume() {
        super.onResume()
        presenter.attachUI(this)

        presenter.setEmail(emailInputText.text.toString())
        presenter.setPassword(passwordInputText.text.toString())
        presenter.setRePassword(rePasswordInputText.text.toString())

        emailInputText.addTextChangedListener(emailTextWatcher)
        passwordInputText.addTextChangedListener(passwordTextWatcher)
        rePasswordInputText.addTextChangedListener(rePasswordTextWatcher)
    }

    override fun onPause() {
        presenter.detachUI()
        emailInputText.removeTextChangedListener(emailTextWatcher)
        passwordInputText.removeTextChangedListener(passwordTextWatcher)
        rePasswordInputText.removeTextChangedListener(rePasswordTextWatcher)
        super.onPause()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putString(EMAIL, emailInputText.text.toString())
        outState.putString(PASSWORD, passwordInputText.text.toString())
        super.onSaveInstanceState(outState)
    }

    override fun setStateButton(isEnabled: Boolean) {
        setButtonStatus(isEnabled)
    }

    override fun setButtonStatus(isEnabled: Boolean) {
        buttonNext.isEnabled = isEnabled
        buttonNext.background = if (isEnabled) ContextCompat.getDrawable(context!!, R.drawable.btn_style_bright)
        else ContextCompat.getDrawable(context!!, R.drawable.btn_disabled)
    }


    override fun showErrorPassword() {
        loadingBar?.visibility = View.GONE

        enableErrorOnPasswordInputs()
        val snackbar = Snackbar.make(view!!, getString(R.string.no_match_password), Snackbar.LENGTH_LONG)
        snackbar.view.setBackgroundColor(Color.RED)
        snackbar.show()
    }


    override fun showErrorEmail() {
        enableErrorOnEmailInput()
        val snackbar = Snackbar.make(view!!, getString(R.string.taken_email_address), Snackbar.LENGTH_LONG)
        snackbar.view.setBackgroundColor(Color.RED)
        snackbar.show()
    }

    override fun showConnectionError() {
        loadingBar?.visibility = View.GONE

        val snackbar = Snackbar.make(view!!, getString(R.string.connection_error), Snackbar.LENGTH_LONG)

        snackbar.view.setBackgroundColor(Color.RED)
        snackbar.show()
    }

    override fun showBackendError() {
        loadingBar?.visibility = View.GONE

        val snackbar = Snackbar.make(view!!, getString(R.string.backend_error), Snackbar.LENGTH_LONG)

        snackbar.view.setBackgroundColor(Color.RED)
        snackbar.show()
    }

    override fun hideKeyboard() {
        hideKeyboard(passwordInputText)
    }

    override fun goToWelcome() {
        //no op
    }

    override fun tagName(): String {
        return SignUpStepTwoFragment::class.java.simpleName
    }



    private val emailTextWatcher = object : TextWatcher {
        override fun afterTextChanged(s: Editable?) {
        }

        override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
        }

        override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
            s?.let { presenter.setEmail(s.toString()) }
            disableErrorOnEmailInput()
        }

    }

    private val passwordTextWatcher = object : TextWatcher {
        override fun afterTextChanged(s: Editable?) {
        }

        override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
        }

        override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
            s?.let { presenter.setPassword(s.toString()) }
            disableErrorOnPasswordInputs()
        }

    }

    private val rePasswordTextWatcher = object : TextWatcher {
        override fun afterTextChanged(s: Editable?) {
        }

        override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
        }

        override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
            s?.let { presenter.setRePassword(s.toString()) }
            disableErrorOnPasswordInputs()
        }

    }

    private fun setUpCredentialsOnRestore(savedInstanceState: Bundle?) {
        val restoredMail = savedInstanceState?.getString(EMAIL)
        val restoredPassword = savedInstanceState?.getString(PASSWORD)

        restoredMail?.let {
            restoredPassword?.let {
                emailInputText.setText(restoredMail)
                passwordInputText.setText(restoredPassword)
                rePasswordInputText.setText(restoredPassword)

                presenter.setEmail(restoredMail)
                presenter.setPassword(restoredPassword)
                presenter.setRePassword(restoredPassword)
            }
        }
    }

    private fun enableErrorOnPasswordInputs() {
        passwordInputText.setBackgroundResource(R.drawable.input_field_error)
        rePasswordInputText.setBackgroundResource(R.drawable.input_field_error)

        isPasswordErrorShown = true
    }

    private fun enableErrorOnEmailInput() {
        emailInputText.setBackgroundResource(R.drawable.input_field_error)

        isEmailErrorShown = true
    }

    private fun disableErrorOnPasswordInputs() {
        if (isPasswordErrorShown) {
            passwordInputText.setBackgroundResource(R.drawable.input_field)
            rePasswordInputText.setBackgroundResource(R.drawable.input_field)
            isPasswordErrorShown = false
        }
    }

    private fun disableErrorOnEmailInput() {
        if (isEmailErrorShown) {
            emailInputText.setBackgroundResource(R.drawable.input_field)
            isEmailErrorShown = false
        }
    }
}
